### ISSUE TITLE
Help wanted: Test Colab notebook for training TFLite models

### ISSUE THREAD

**Issue by EdjeElectronics** [EdjeElectronics/TensorFlow-Lite-Object-Detection-on-Android-and-Raspberry-Pi#135]
It's been a while since I've worked on this repository, but I'm diving back into it to make some improvements! Today I added a Google Colab notebook that allows you use Google's servers to train, convert, test, and export a TensorFlow Lite model. It makes training a custom TFLite detection model easy! It uses the TensorFlow Object Detection API, which provides high configurability and is best for training with large datasets.

I'm looking for help with testing the Colab notebook to confirm it works for all users. Have you been looking to train an SSD-MobileNet or EfficientDet model and and deploy it with TFLite? If so, can you try stepping through this notebook to see if it works for you and post in this thread on how it turned out? Let me know if you run in to any errors or have questions.

Here's a link to the Colab notebook: [Train_TFLite2_Object_Detction_Model.ipynb](./Train_TFLite2_Object_Detction_Model.ipynb)

<a href="https://colab.research.google.com/github/EdjeElectronics/TensorFlow-Lite-Object-Detection-on-Android-and-Raspberry-Pi/blob/master/Train_TFLite2_Object_Detction_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

If you need a dataset to use, I uploaded my ["bird, squirrel, raccoon" dataset of 900 images](https://www.dropbox.com/s/en33x280e4z3wbt/BSR2.zip?dl=0) to Dropbox. I included a command to download this dataset into the Colab as one of the options in Step 3.

I'm planning to make a video that shows how to go through the notebook step-by-step, but it will be a few weeks until it's ready. For now, hopefully the instructions inside the Colab doc do a good enough job explaining.

**Feedback I'm looking for:**

- Were you able to successfully train a ssd-mobilenet-v2 model, test it, and see detection results on the test images? (i.e. were you able to make it through Steps 1 - 7 without errors?)
- If you ran into errors, what were they?
- Was anything in the notebook confusing or unclear? Were there any points where you weren't sure what to do next?
- Do you have any suggestions for improving the notebook?

**Known issues:**
- The `centernet-mobilenet` model still doesn't work with TensorFlow Lite, I'm still trying to get that one figured out.
- Accuracy drops significantly when quantizing the ssd-mobilenet-v2 model
- efficentdet, centernet-mobilenet, and ssd-mobilenet-fpnlite models can't be quantized
- In other words, quantization still isn't really working. But the unquantized TFLite models seem to work well!

**Comment by elvenkim1** [comment]
Hi Edje

I have this error that says bird is not in the list. I downloaded your sample bird list 
![image](https://user-images.githubusercontent.com/88960981/188524177-c1e8e644-8360-4946-a04f-c7e2bf624f45.png)

**Comment by EdjeElectronics** [comment]
@elvenkim1 it looks like you deleted your previous comment which showed the error. Can you show me the error you're getting again?

Thanks!

**Comment by elvenkim1** [comment]
![image](https://user-images.githubusercontent.com/88960981/188809827-1735c5ec-863f-4268-87b8-d53e73c416e7.png)

I am able to run now and completed all steps without errors.

I tried with my own dataset and bumped into this error. Please advise.

**Comment by elvenkim1** [comment]
also, there is missing on label on the image itself. 
![image](https://user-images.githubusercontent.com/88960981/188814137-72625ba3-9b87-4a49-af39-a3c9a60e07d5.png)

**Comment by elvenkim1** [comment]
![image](https://user-images.githubusercontent.com/88960981/188833690-dff0af74-7cb2-492c-8e76-d8bc562f379a.png)

the video is not showing the label as well. I tested this on Raspberry Pi

python TFLite_detection_webcam.py --imgdir=fine_tuned_model_lite  --> should be python TFLite_detection_webcam.py --modeldir=fine_tuned_model_lite

**Comment by elvenkim1** [comment]
Hi EJ

yes, it is working now 👍! You are right to say the model is not accurate..as I only trained for 500 steps. 

It should be: thresh --> threshold

TFLite_detection_webcam.py --modeldir=fine_tuned_model_lite --threshold=0.01

Now may I know if you know how to add the accuracy, precision etc graph? I understand we need to have another terminal running alongside with the training process but this is not possible with Colab.

**Comment by EdjeElectronics** [comment]
Awesome! Glad it's working. The TensorBoard window in Section 5 will show the model's training and validation loss as it trains. Unfortunately, it's tough to show a graph of precision/accuracy over time. I will add a section to calculate the model's mAP on the test images.

**Comment by elvenkim1** [comment]
Awesome, thanks!

**Comment by kleinpoe** [comment]
I went through the board and am now in the training stage up to which it worked beautifully. I have used your bird, squirrel, racoon dataset, as I happen to be interested in detecting the squirrels and birds in my garden :-D
A minor thing I have noticed:
"At this point, whether you used Option 1 or Option 2, you should be able to click the folder icon..." Should also include "Option 3"

I think it would be helpful to post some information on your reference dataset: it would be interesting to compare the obtained score with yours and maybe have some example video to try it out with in the end to have a confirmation that it worked. Maybe you could also provide the tflite file. (I don't know if its helpful for anybody, but as I am interested in exactly these objects at least I am very interested :-) )

In sum, this is an awesome project you have there, thank you very much

**Comment by RobotGrrl** [comment]
At step # "Next, we'll split the images into train, validation, and test sets. Here's what each set is used for:"

The [python script that is executed](https://raw.githubusercontent.com/EdjeElectronics/TensorFlow-Lite-Object-Detection-on-Android-and-Raspberry-Pi/master/util_scripts/train_val_test_split.py) uses the wrong directory, making the total image count 0. The script looks in /content/images/all, but the path of the images are /content/images/all/images (based on following the steps in the Colab notebook).

**Comment by RobotGrrl** [comment]
On Step #23 (? where are the step numbers anyway?)

"NotFoundError: /content/labelmap.pbtxt; No such file or directory"

This is because the file is named "labelmap.txt", not ".pbtxt"

**Comment by RobotGrrl** [comment]
After fixing the above, a new error is reported on Step # 23

```
ParseError: 1:1 : Message type "object_detection.protos.StringIntLabelMap" has no field named "class1".


During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)

[/usr/local/lib/python3.7/dist-packages/object_detection/utils/label_map_util.py](https://localhost:8080/#) in load_labelmap(path)
    171       text_format.Merge(label_map_string, label_map)
    172     except text_format.ParseError:
--> 173       label_map.ParseFromString(label_map_string)
    174   _validate_label_map(label_map)
    175   return label_map

TypeError: a bytes-like object is required, not 'str'
```

**Comment by RobotGrrl** [comment]
At Step # 9 received this error:

```
Successfully converted xml to csv.
Successfully converted xml to csv.
Traceback (most recent call last):
  File "create_tfrecord.py", line 120, in <module>
    tf.app.run()
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/platform/app.py", line 36, in run
    _run(main=main, argv=argv, flags_parser=_parse_flags_tolerate_undef)
  File "/usr/local/lib/python3.7/dist-packages/absl/app.py", line 308, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.7/dist-packages/absl/app.py", line 254, in _run_main
    sys.exit(main(argv))
  File "create_tfrecord.py", line 98, in main
    tf_example = create_tf_example(group, path)
  File "create_tfrecord.py", line 46, in create_tf_example
    encoded_jpg = fid.read()
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/lib/io/file_io.py", line 114, in read
    self._preread_check()
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/lib/io/file_io.py", line 77, in _preread_check
    compat.path_to_str(self.__name), 1024 * 512)
tensorflow.python.framework.errors_impl.NotFoundError: /content/images/train/image-01.jpeg; No such file or directory
Traceback (most recent call last):
  File "create_tfrecord.py", line 120, in <module>
    tf.app.run()
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/platform/app.py", line 36, in run
    _run(main=main, argv=argv, flags_parser=_parse_flags_tolerate_undef)
  File "/usr/local/lib/python3.7/dist-packages/absl/app.py", line 308, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.7/dist-packages/absl/app.py", line 254, in _run_main
    sys.exit(main(argv))
  File "create_tfrecord.py", line 98, in main
    tf_example = create_tf_example(group, path)
  File "create_tfrecord.py", line 46, in create_tf_example
    encoded_jpg = fid.read()
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/lib/io/file_io.py", line 114, in read
    self._preread_check()
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/lib/io/file_io.py", line 77, in _preread_check
    compat.path_to_str(self.__name), 1024 * 512)
tensorflow.python.framework.errors_impl.NotFoundError: /content/images/validation/image-01.jpg; No such file or directory
```

There is no image-01.jpg in the uploaded images.zip folder. There is a image-001.jpg though, but that is located in /train

**Comment by RobotGrrl** [comment]
> At Step # 9 received this error:
> There is no image-01.jpg in the uploaded images.zip folder. There is a image-001.jpg though, but that is located in /train

Sorry, this issue is because the filenames inside the .xml from LabelImg are different than the actual filenames of the .jpg files.

**Comment by RobotGrrl** [comment]
On Step # 44 (run training), receive this error:

```
Traceback (most recent call last):
  File "/content/models/research/object_detection/model_main_tf2.py", line 31, in <module>
    from object_detection import model_lib_v2
  File "/usr/local/lib/python3.7/dist-packages/object_detection/model_lib_v2.py", line 31, in <module>
    from object_detection import model_lib
  File "/usr/local/lib/python3.7/dist-packages/object_detection/model_lib.py", line 35, in <module>
    from object_detection.builders import optimizer_builder
  File "/usr/local/lib/python3.7/dist-packages/object_detection/builders/optimizer_builder.py", line 25, in <module>
    from official.modeling.optimization import ema_optimizer
  File "/usr/local/lib/python3.7/dist-packages/official/modeling/optimization/__init__.py", line 23, in <module>
    from official.modeling.optimization.optimizer_factory import OptimizerFactory
  File "/usr/local/lib/python3.7/dist-packages/official/modeling/optimization/optimizer_factory.py", line 36, in <module>
    'adamw_experimental': tf.keras.optimizers.experimental.AdamW,
AttributeError: module 'tensorflow.keras.optimizers' has no attribute 'experimental'
```

**Comment by RobotGrrl** [comment]
On Step # 43 (loading tensorboard), a 403 Error is displayed:

<img width="959" alt="Screen Shot 2022-09-27 at 10 03 23 PM" src="https://user-images.githubusercontent.com/190496/192625142-0cf46c1e-f28d-4098-aba1-2e8b1883d71f.png">

**Comment by jugglingboss** [comment]
I received the same error as @RobotGrrl 

![image](https://user-images.githubusercontent.com/40720328/192655303-05f782c7-5195-4c93-8906-f89fdacce766.png)

**Comment by jugglingboss** [comment]
@RobotGrrl I got past this error by editing /usr/local/lib/python3.7/dist-packages/official/modeling/optimization/optimizer_factory.py and removing line 36,    

 'adamw_experimental': tf.keras.optimizers.experimental.AdamW,

**Comment by EdjeElectronics** [comment]
Thanks for the help and feedback everyone!
